File: lab7/Dictionary.java
//-----------------------------------------------------------------------------
// Dictionary.java
// Binary Search Tree implementation of the Dictionary ADT
//-----------------------------------------------------------------------------

public class Dictionary implements DictionaryInterface{
  
  // private inner Node class
  private class Node {
    String key;
    String value;
    Node left;
    Node right;
    
    Node(String key, String value){
      this.key = key;
      this.value = value;
      left = null;
      right = null;
    }
  }
  
  // Fields for the Dictionary class
  private Node root;     // reference to first Node in List
  private int numItems;  // number of items in this IntegerList
  
  // IntegerList()
  // constructor for the IntegerList class
  public Dictionary(){
    root = null;
    numItems = 0;
  }
  
  
  
  // findKey()
// returns the Node containing key k in the subtree rooted at R, or returns
// NULL if no such Node exists
  public Node findKey(Node R, String key){
    if(R==null || key.compareTo(R.key)==0)
      return R;
    if(key.compareTo(R.key)<0){
      return findKey(R.left, key);
    }else
      return findKey(R.right, key);
  }
  
  
  
// findParent()
// returns the parent of N in the subtree rooted at R, or returns NULL 
// if N is equal to R. (pre: R != NULL)
  Node findParent(Node N, Node R){
    Node P=null;
    if( N!=R ){
      P = R;
      while( P.left!=N && P.right!=N ){
        if((N.key).compareTo(P.key)<0)
          P = P.left;
        else
          P = P.right;
      }
    }
    return P;
  }
  
  // findLeftmost()
// returns the leftmost Node in the subtree rooted at R, or NULL if R is NULL.
  Node findLeftmost(Node R){
    Node L = R;
    if( L!=null ) for( ; L.left!=null; L=L.left) ;
    return L;
  }
  
// printInOrder()
// prints the (key, value) pairs belonging to the subtree rooted at R in order
// of increasing keys to file pointed to by out.
 void printInOrder(Node R){
    String s;
    if( R!=null ){
      printInOrder(R.left);
      System.out.println(R.key +" " + R.value);
      printInOrder(R.right);
    }
  }
  
  
  // isEmpty()
  // returns true if this Dictionary is empty, false otherwise
  // pre: none
  public boolean isEmpty(){
    return (numItems==0);
  }
  
  // size()
  // returns the number of entries in this Dictionary
  // pre: none
  public int size(){
    return numItems;
  }
  
  // lookup()
  // returns value associated key, or null reference if no such key exists
  // pre: none
  public String lookup(String key){
    Node N = findKey(root, key);
    return ( N==null ? null : N.value );
  }
  
  // insert()
  // inserts new (key,value) pair into this Dictionary
  // pre: lookup(key)==null
  public void insert(String key, String value) throws KeyCollisionException{
    if(findKey(root, key) != null){
      throw new KeyCollisionException(
                                      "Dictionary Error: cannot insert() duplicate key:" + key);
    }
    Node N = new Node(key, value);
    Node B = null;
    Node A = root;
    while( A!=null ){
      B = A;
      if(key.compareTo(A.key)<0 ) A = A.left;
      else A = A.right;
    }
    if( B==null ) root = N;
    else if(key.compareTo(B.key)<0 ) B.left = N;
    else B.right = N;
    numItems++;
  }
  
  
  // delete()
  // deletes pair with the given key
  // pre: lookup(key)!=null
  public void delete(String key) throws KeyNotFoundException{
    if( lookup(key) == null ){  
      throw new KeyNotFoundException("cannot delete non-existent key");
    }
    Node N, P, S;
    N = findKey(root, key);
    if( N.left==null && N.right==null ){
      if( N==root ){
        root = null;
      }else{
        P = findParent(N, root);
        if( P.right==N ) P.right = null;
        else P.left = null;
      }
    }else if( N.right==null ){
      if( N==root ){
        root = N.left;
      }else{
        P = findParent(N, root);
        if( P.right==N ) P.right = N.left;
        else P.left = N.left;
      }
    }else if( N.left==null ){
      if( N==root ){
        root = N.right;
      }else{
        P = findParent(N, root);
        if( P.right==N ) P.right = N.right;
        else P.left = N.right;
      }
    }else{  // N->left!=NULL && N->right!=NULL
      S = findLeftmost(N.right);
      N.key = S.key;
      N.value = S.value;
      P = findParent(S, N);
      if( P.right==S ) P.right = S.right;
      else P.left = S.right;
    }
    numItems--;
  }
  
  
  
  // makeEmpty()
  // pre: none
  public void makeEmpty(){
    root = null;
    numItems = 0;
  }
  
  // toString()
  // returns a String representation of this Dictionary
  // overrides Object's toString() method
  // pre: none
  public String toString(){
    printInOrder(root);
    return " ";
  }
}

File: lab5/Dictionary.c
//-----------------------------------------------------------------------------
//   IntegerStack.c
//   Implementation file for IntegerStack ADT
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"


// private types --------------------------------------------------------------

// NodeObj
typedef struct NodeObj{
   char* key;
   char* value;
   struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* x, char* y) {
   Node N = malloc(sizeof(NodeObj));
   assert(N!=NULL);
   N->key = x;
   N->value = y;
   N->next = NULL;
   return(N);
}

// freeNode()
// destructor for the Node type
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}


// DictionaryObj
typedef struct DictionaryObj{
   Node top;
   int numItems;
} DictionaryObj;


// public functions -----------------------------------------------------------


//findKey()
//retruns a node with a specific key
Node findKey(Dictionary D, char* k){
   if( D==NULL ){
   fprintf(stderr, 
           "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
   exit(EXIT_FAILURE);
   }

   Node N = D->top;
   while(strcmp(N->key, k)){
      N = N->next;
   }
   return N;
}



// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void){
   Dictionary D = malloc(sizeof(DictionaryObj));
   assert(D!=NULL);
   D->top = NULL;
   D->numItems = 0;
   return D;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
   if( pD!=NULL && *pD!=NULL ){
      if( !isEmpty(*pD) ) makeEmpty(*pD); //empty the Dictionary
      free(*pD);
      *pD = NULL;
   }
}

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){
   if( D==NULL ){
      fprintf(stderr, 
              "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   return(D->numItems==0);
}

// size()
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D){
   f( D==NULL ){
      fprintf(stderr, 
              "Dictionary Error: calling size() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   return (D->numItems);
}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no 
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* k){ 
if( D==NULL ){
      fprintf(stderr, 
              "Dictionary Error: calling lookup() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
Node N; 
char* val = NULL;
for(N=D->top; N!=NULL; N=N->next)
   if(strcmp(N->key, k)) val = N->value;
return val;
}

// push()
// pushes x onto top of S
// pre: none
void insert(Dictionary D, char* x, char* y){
   Node N;
   if( D==NULL ){
      fprintf(stderr, "Stack Error: calling push() on NULL Stack reference\n");
      exit(EXIT_FAILURE);
   }
   N = newNode(x,y);
   N->next = D->top;
   D->top = N;
   D->numItems++;
}

// delete()
// deletes and returns item at top of D
// pre: !isEmpty(S)
void delete(Stack S, char* k){
   int returnValue;
   Node curr;
   Node N;
   if( S==NULL ){
      fprintf(stderr, "Stack Error: calling delete() on NULL Dictionary reference\n");
      exit(EXIT_FAILURE);
   }
   if( S->numItems==0 ){
      fprintf(stderr, "Stack Error: calling delete() on empty Dictionary\n");
      exit(EXIT_FAILURE);
   }

   N = D->top;
   returnValue = D->top->item;
   D->top = D->top->next;
   D->numItems--;
   freeNode(&N);
   return returnValue;
}


// makeEmpty()
// re-sets D to the empty state.
// pre: none
void makeEmpty(Dictionary D){
   D->top = NULL;
   D->numItems = 0;
   freeNode(&D);
}

// printStack()
// prints a text representation of S to the file pointed to by out
// pre: none
void printStack(FILE* out, Dictionary D){
   Node N;
   if( D==NULL ){
      fprintf(stderr, 
              "Stack Error: calling printStack() on NULL Stack reference\n");
      exit(EXIT_FAILURE);
   }
   for(N=D->top; N!=NULL; N=N->next) fprintf(out, "%d ", N->item);
   fprintf(out, "\n");
}

